src/System.Collections/tests/Generic/Comparers/Comparer.Tests.cs

```diff
@@ -0,0 +1,321 @@
+        [MemberData(nameof(UIntEnumComparisonsData))]
+        [MemberData(nameof(LongEnumComparisonsData))]
+        [MemberData(nameof(PlainObjectComparisonsData))]
+        public void MostComparisons<T>(T left, T right, int expected) // xUnit is awesome and supports generic theories :-)
+        {
+            var comparer = Comparer<T>.Default;
+            Assert.Equal(expected, Math.Sign(comparer.Compare(left, right)));
+            Assert.Equal(0, comparer.Compare(left, left));
+            Assert.Equal(0, comparer.Compare(right, right));
+
+            IComparer nonGenericComparer = comparer;
+            // If both sides are Ts then the explicit implementation of IComparer.Compare
+            // should also succeed, with the same results
+            Assert.Equal(expected, Math.Sign(nonGenericComparer.Compare(left, right)));
+            Assert.Equal(0, nonGenericComparer.Compare(left, left));
+            Assert.Equal(0, nonGenericComparer.Compare(right, right));
+
+            // All comparers returned by Comparer<T>.Default should be able
+            // to handle nulls before dispatching to IComparable<T>.CompareTo()
+            if (default(T) == null) // This will be true if T is a reference type or nullable
+            {
+                T nil = default(T);
+                Assert.Equal(0, comparer.Compare(nil, nil));
+
+                // null should be ordered before/equal to everything (never after)
+                // We assert that it's -1 or 0 in case left/right is null, as well
+                // We assert that it's -1 rather than any negative number since these
+                // values are hardcoded in the comparer logic, rather than being left
+                // to the object being compared
+                Assert.InRange(comparer.Compare(nil, left), -1, 0);
+                Assert.InRange(comparer.Compare(nil, right), -1, 0);
+
+                Assert.InRange(comparer.Compare(left, nil), 0, 1);
+                Assert.InRange(comparer.Compare(right, nil), 0, 1);
+
+                // Validate behavior for the IComparer.Compare implementation, as well
+                Assert.Equal(0, nonGenericComparer.Compare(nil, nil));
+
+                Assert.InRange(nonGenericComparer.Compare(nil, left), -1, 0);
+                Assert.InRange(nonGenericComparer.Compare(nil, right), -1, 0);
+
+                Assert.InRange(nonGenericComparer.Compare(left, nil), 0, 1);
+                Assert.InRange(nonGenericComparer.Compare(right, nil), 0, 1);
+            }
+        }
+
+        public static IEnumerable<object[]> IComparableComparisonsData()
+        {
+            var testCases = new[]
+            {
+                Tuple.Create(new GenericComparable(3), new GenericComparable(4), -1),
+                Tuple.Create(new GenericComparable(5), new GenericComparable(2), 1),
+                Tuple.Create(new GenericComparable(int.MinValue), new GenericComparable(int.MinValue), 0),
```

Comment:
When the values are the same, they'll be the same in reverse... we don't need to run the same test twice. (Same for other occurrences of this below.)